### PR TITLE
CR-1106861 : xbutil/xbmgmt: Forego device specification for single device systems

### PR DESCRIPTION
**Pull request by rozumx**
At the top program option parser, the optional options --device & -d were added with default and implicit values.  If the user specifies this the 'device' option with no value and there is only 1 device installed, it will be selected and passed to the sub-command.

Work Done
---------
+ Added optional --device & -d option with both default (e.g. "") and implicit (e.g., "default") values to the top-level program option tree.
+ Updated XBMain::_main() to automatically select a device if the is only one device and the users specifies the device option.
+ Added DRC check for systems with no and multiple devices installed.
+ Cleaned up the SubCmdProgram::execute() method.

### CODE REVIEW

**Comment by xilinxgitops** [comment]
retest this please

**Comment by xilinxgitops** [comment]
retest this please

**Comment by xilinxgitops** [comment]
retest this please

**Comment by rozumx** [comment]
retest this please

File: src/runtime_src/core/tools/common/XBMain.cpp [hunk 1]
```diff
@@ -145,6 +149,41 @@ void  main_(int argc, char** argv,
...
+    sDevice.clear();
+    boost::property_tree::ptree available_devices = XBU::get_available_devices(isUserDomain);
+
+    if (available_devices.empty()) 
+      throw std::runtime_error("No devices found.");
+
+    // If there are multiple devices, then no default device can be found.
+    if (available_devices.size() > 1) {
+      std::cerr << "\nERROR: Multiple devices found. Please specify a single device using the --device option\n\n";
+      std::cout << "List of available devices:" << std::endl;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/XBMain.cpp:164]:
```suggestion
      std::cerr << "List of available devices:" << std::endl;

```

File: src/runtime_src/core/tools/common/XBMain.cpp [hunk 2]
```diff
@@ -145,6 +149,41 @@ void  main_(int argc, char** argv,
...
+      for (auto &kd : available_devices) {
+        boost::property_tree::ptree& _dev = kd.second;
+        std::cout << boost::format("  [%s] : %s\n") % _dev.get<std::string>("bdf") % _dev.get<std::string>("vbnv");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/XBMain.cpp:167]:
```suggestion
        std::cerr << boost::format("  [%s] : %s\n") % _dev.get<std::string>("bdf") % _dev.get<std::string>("vbnv");

```

File: src/runtime_src/core/tools/common/XBMain.cpp [hunk 3]
```diff
@@ -145,6 +149,41 @@ void  main_(int argc, char** argv,
...
+
+    if (available_devices.empty()) 
+      throw std::runtime_error("No devices found.");
+
+    // If there are multiple devices, then no default device can be found.
+    if (available_devices.size() > 1) {
+      std::cerr << "\nERROR: Multiple devices found. Please specify a single device using the --device option\n\n";
+      std::cout << "List of available devices:" << std::endl;
+      for (auto &kd : available_devices) {
+        boost::property_tree::ptree& _dev = kd.second;
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/XBMain.cpp:166]:
```suggestion
        boost::property_tree::ptree& dev = kd.second;

```

File: src/runtime_src/core/tools/common/XBMain.cpp [hunk 4]
```diff
@@ -66,6 +69,7 @@ void  main_(int argc, char** argv,
   // Hidden Options
   po::options_description hiddenOptions("Hidden Options");
   hiddenOptions.add_options()
+    ("device,d",    boost::program_options::value<decltype(sDevice)>(&sDevice)->default_value("")->implicit_value("default"), "If specified with no BDF value and there is only 1 device, that device will be automatically selected.\n")
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/XBMain.cpp:72]: This addresses
`xbutil examine -d`
use case?

**Review by AShivangi — CHANGES_REQUESTED** [review]
This looks good, but we need to propagate these changes to all the subcommands in b=xbutil and xbmgmt

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 4 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp:72]: Please see existing PR comments.

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 3 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp]: done

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 2 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp]: done

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 3 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp]: done

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 2 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp]: done

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 1 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp]: done

File: src/runtime_src/core/tools/common/XBMain.cpp (hunk 4 above)
Comment (rozumx) [reply src/runtime_src/core/tools/common/XBMain.cpp:72]: Please see existing submission description.  First paragraph.

**Review by AShivangi — APPROVED** [review]
Approving this PR for now. We will be removing dead code for parsing device entries soon.
Thanks for making these changes

**Pull request merged**